src/model/CoilHeatingWaterToAirHeatPumpEquationFit.cpp

```diff
@@ -304,22 +306,22 @@ namespace model {
       return boost::none;
     }
 
-    CurveQuadLinear CoilHeatingWaterToAirHeatPumpEquationFit_Impl::heatingCapacityCurve() const {
+    Curve CoilHeatingWaterToAirHeatPumpEquationFit_Impl::heatingCapacityCurve() const {
       WorkspaceObject wo = getTarget(OS_Coil_Heating_WaterToAirHeatPump_EquationFitFields::HeatingCapacityCurveName).get();
-      return wo.cast<CurveQuadLinear>();
+      return wo.cast<Curve>();
     }
 
-    bool CoilHeatingWaterToAirHeatPumpEquationFit_Impl::setHeatingCapacityCurve(const CurveQuadLinear& heatingCapacityCurve) {
+    bool CoilHeatingWaterToAirHeatPumpEquationFit_Impl::setHeatingCapacityCurve(const Curve& heatingCapacityCurve) {
```

Comment:
This specifically expects a `QuadvariateFunctions` one. Should we check that `heatingCapacityCurve.numVariables() == 4` here?

Genuine question. Not saying we should necessarily, but since this is the general move of E+ to allow generic curve and table objects everywhere, this is bound to come up again, I think now's a good time to decide what to do/standardize. @kbenne @joseph-robertson 

**Pros:**

* Checking is good because it catches user errors early instead of letting E+ complain (via `CheckCurveDim`)

**Cons:**

* More work for us
* Options:
    * Option1: we are hardcoding a `curve.numVariables()` check in each setter that takes a generic `Curve`, which potentially could fall out of date (though in this case I don't see why the number of expected independent variables would change suddenly in E+)
    * Option 2: The alternative is to add a special case in `WorkspaceObject_Impl::setPointer` or a new method `WorkspaceObject_Impl::setCurvePointer`
        * A new check in `setPointer` is going to impose a performance penalty for all objects
        * A new method specifically for Curves means we gotta be careful to use this one specifically for curves

Reply:
actually for option 2, the best option is probably to a new `bool ModelObject_Impl::setCurve(unsigned index, const Curve& curve)` method that would check the IDD references and do the numVariables() check for us before setting the handle via `WorkspaceObject_Impl::setPointer` (or `WorkspaceObject_Impl::setString` for speed).

That'd avoid having to do lots of shenanigans in Workspace land to figure out the number of variables a Table object has.

Reply:
Increasingly, I'm starting to appreciate some value in not making OpenStudio too smart and therefore letting EnergyPlus report certain errors. 

That said if we do go with validation in OpenStudio then option 2, but with a new ::setCurve method seems right. Paying the performance penality for every call to the existing ::setPointer seems unacceptable to me.

Reply:
I'm opening an issue here, that we may just end up closing. I realize the Curve objects are already fine (via the `\reference` check), only the TableLookup has this issue:

* https://github.com/NREL/OpenStudio/issues/4982